src/virtio/virtio_net_emul.c

```diff
@@ -0,0 +1,363 @@
+
+    /* grab the next receive chain */
+    uint16_t guest_idx = vring->avail->idx;
+    uint16_t idx = vqs[RX_QUEUE].last_idx;
+
+    if (idx == guest_idx) {
+        printf("\"%s\"|VIRTIO NET|WARNING: queue is full, drop the packet\n", sel4cp_name);
+        return 1;
+    }
+
+    struct virtio_net_hdr_mrg_rxbuf virtio_hdr;
+    memzero(&virtio_hdr, sizeof(virtio_hdr));
+
+    /* total length of the copied packet */
+    size_t copied = 0;
+    /* amount of the current descriptor copied */
+    size_t desc_copied = 0;
+
+    /* read the head of the descriptor chain */
+    uint16_t desc_head = vring->avail->ring[idx % vring->num];
+    uint16_t curr_desc_head = desc_head;
+
+    // have we finished copying the net header?
+    bool net_header_processed = false;
+
+    do {
+        /* determine how much we can copy */
+        uint32_t copying;
+        /* what are we copying? */
+        void *buf_base = NULL;
+
+        // process the net header
+        if (!net_header_processed) {
+            copying = sizeof(struct virtio_net_hdr_mrg_rxbuf) - copied;
+            buf_base = &virtio_hdr;
+
+        // otherwise, process the actual packet
+        } else {
+            copying = size - copied;
+            buf_base = buf;
+        }
+
+        copying = MIN(copying, vring->desc[curr_desc_head].len - desc_copied);
+
+        memcpy((void *)vring->desc[curr_desc_head].addr + desc_copied, buf_base + copied, copying);
+
+        /* update amounts */
+        copied += copying;
+        desc_copied += copying;
+
+        // do we need another buffer from the virtqueue?
+        if (desc_copied == vring->desc[curr_desc_head].len) {
+            if (!vring->desc[curr_desc_head].flags & VRING_DESC_F_NEXT) {
+                /* descriptor chain is too short to hold the whole packet.
+                * just truncate */
+                break;
+            }
+            curr_desc_head = vring->desc[curr_desc_head].next;
+            desc_copied = 0;
+        }
+
+        // have we finished copying the net header?
+        if (copied == sizeof(struct virtio_net_hdr_mrg_rxbuf)) {
+            copied = 0;
+            net_header_processed = true;
+        }
+
+    } while (!net_header_processed || copied < size);
+
+    // record the real amount we have copied
+    if (net_header_processed) {
+        copied += sizeof(struct virtio_net_hdr_mrg_rxbuf);
+    }
+    /* now put it in the used ring */
+    struct vring_used_elem used_elem = {desc_head, copied};
+    uint16_t used_idx = vring->used->idx;
+
+    vring->used->ring[used_idx % vring->num] = used_elem;
+    vring->used->idx++;
+
+    /* record that we've used this descriptor chain now */
+    vqs[RX_QUEUE].last_idx++;
+
+    // set the reason of the irq
+    struct virtio_emul_handler *handler = get_virtio_net_emul_handler();
+    assert(handler);
+    handler->data.InterruptStatus = BIT_LOW(0);
+
+    // notify the guest
+    bool success = vgic_inject_irq(VCPU_ID, VIRTIO_NET_IRQ);
```

Comment:
Not necessarily to be fixed here but a note to myself that we can't have any kind of architecture specific stuff in the virtIO code. So when we need to deliver an interrupt to the guest, we need an arch-independent way of doing it.

Reply:
@abrandnewusername if you could point out the other potentially arch-specific bits that would be really helpful